### PR TITLE
Retire the reshuffle method for table data expansion

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
This method was introduced to improve the data redistribution
performance during gpexpand phase2, however per benchmark results the
effect does not reach our expectation.  For example when expanding a
table from 7 segments to 8 segments the reshuffle method is only 30%
faster than the traditional CTAS method, when expanding from 4 to 8
segments reshuffle is even 10% slower than CTAS.  When there are indexes
on the table the reshuffle performance can be worse, and extra VACUUM is
needed to actually free the disk space.  According to our experiments
the bottleneck of reshuffle method is on the tuple deletion operation,
it is much slower than the insertion operation used by CTAS.

The reshuffle method does have some benefits, it requires less extra
disk space, it also requires less network bandwidth (similar to CTAS
method with the new JCH reduce method, but less than CTAS + MOD).  And
it can be faster in some cases, however as we can not automatically
determine when it is faster it is not easy to get benefit from it in
practice.

On the other side the reshuffle method is less tested, it is possible to
have bugs in corner cases, so it is not production ready yet.

In such a case we decided to retire it entirely for now, we might add it
back in the future if we can get rid of the slow deletion or find out
reliable ways to automatically choose between reshuffle and ctas
methods.

Discussion: https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/8xknWag-SkI/5OsIhZWdDgAJ

## Here are some reminders before you submit the pull request
- [x] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/expected/expand_table.out [hunk 1]
```diff
@@ -635,7 +647,7 @@ select gp_segment_id, * from mix_base_tbl order by 2, 1;
              1 | 12 | 11
              2 | 13 | 12
              2 | 14 | 13
-             0 | 22 | 21
+             2 | 22 | 21
```
Comment (hlinnaka) [inline src/test/regress/expected/expand_table.out]: Why did this change?

File: src/test/regress/expected/gangsize.out [hunk 2]
```diff
@@ -231,7 +231,7 @@ INFO:  Distributed transaction command 'Distributed Abort (No Prepared)' to ALL
 begin;
 alter table replicate_2_1 expand table;
 INFO:  (slice 0) Dispatch command to ALL contents: 0 1 2
-INFO:  (slice 1) Dispatch command to PARTIAL contents: 0 1
+INFO:  (slice 1) Dispatch command to SINGLE content
```
Comment (hlinnaka) [inline src/test/regress/expected/gangsize.out]: What's up with this?

**Comment by ashwinstar** [comment]
Might we worth mentioning for future and to keep everyone on same page, even without reshuffle method the advantage consistent hash continues to provide for redistribution during gpexpand.

File: src/test/regress/expected/gangsize.out (hunk 2 above)
Comment (pivotal-ning-yu) [reply src/test/regress/expected/gangsize.out]: The CTAS method uses a different way with the reshuffle method to redistribute a replicated table.

For reshuffle method all the old segments are involved during the redistribution, in this case segment 0 sends data to segment 0 and 2, segment 1 sends data to segment 0, so slice 1 contains two contents.

For CTAS method data are copied from only one old segment, so slice 1 contains only one content.

File: src/test/regress/expected/expand_table.out (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/test/regress/expected/expand_table.out]: This was actually a bug in the reshuffle method.  For an inherited table the reshuffle method will always pass root distribution policy to the children, so when root and children tables have different policies the result can be wrong.  In the past we fixed a similar issue that root numsegments is passed to all the children, later when we noticed this new issue we already decided to retire the reshuffle method so we didn't push fix for the issue.

**Comment by pivotal-ning-yu** [comment]
> Might we worth mentioning for future and to keep everyone on same page, even without reshuffle method the advantage consistent hash continues to provide for redistribution during gpexpand.

Yes, the CTAS performance is also benefited from the JCH reduce method, less data are transferred across segments.  Suppose a table has K rows of records, when redistributing it from M segments to N segments using the CTAS method (N > M):
- using legacy MOD reduce method:
  - each old segment: transfer out (N-1)K/MN, transfer in (M - 1)K/MN
  - each new segment: transfer in K/N
- using new JCH reduce method:
  - each old segment: transfer out (N-M)K/MN
  - each new segment: transfer in K/N

NOTE: here we only count the data transferred __in__/__out__ across segments via network.  All the data still need to be transferred, however the data that transferred back to the same segment will be optimized by the kernel to go through the loopback interface which is must faster than external network.

We can see that JCH helps to reduce data transfer significantly on old segments, this makes it faster, and it also helps to reduce the bandwidth usage so in theory concurrent queries on other tables can also have better performance.

**Comment by ashwinstar** [comment]
PR is missing to provide clear justification of why the reshuffle method came out as looser.  Wish to understand the retirement is due to performance reasons or implementation risk. Tried to search mailing list but failed to find clear rationals for this there either.

Are we retiring reshuffle just to avoid the risk of new implementation and continue to use the tried and tested method? or ignoring that fact benchmarks proved its a concept worth not visiting at all in future if we had no time pressure?

Please can you post benchmarks numbers to clarify why reshuffle method didn't help. General cases of expansion would be going from 100 segments to 110 maybe, doesn't it prove to be not helpful for such cases? Reading full data and hash calculation for all the visible tuples is consistent between reshuffle and ctas. Reshuffle adds marking buffer dirty for deletes which ctas doesn't but ctas always has full rewrite of data. But for AO / CO tables reshuffle might still be winner over ctas as only should be touching visimap. Understand reshuffle can generate more work for future vacuum but still should help to get expand performed faster which was one of the goals. Where exactly is the bottleneck for reshuffle?

**Comment by pivotal-ning-yu** [comment]
> PR is missing to provide clear justification of why the reshuffle method came out as looser. Wish to understand the retirement is due to performance reasons or implementation risk. Tried to search mailing list but failed to find clear rationals for this there either.
> 
> Are we retiring reshuffle just to avoid the risk of new implementation and continue to use the tried and tested method? or ignoring that fact benchmarks proved its a concept worth not visiting at all in future if we had no time pressure?
> 
> Please can you post benchmarks numbers to clarify why reshuffle method didn't help. General cases of expansion would be going from 100 segments to 110 maybe, doesn't it prove to be not helpful for such cases? Reading full data and hash calculation for all the visible tuples is consistent between reshuffle and ctas. Reshuffle adds marking buffer dirty for deletes which ctas doesn't but ctas always has full rewrite of data. But for AO / CO tables reshuffle might still be winner over ctas as only should be touching visimap. Understand reshuffle can generate more work for future vacuum but still should help to get expand performed faster which was one of the goals. Where exactly is the bottleneck for reshuffle?

@ashwinstar previous discussion and some benchmark data can be found here: https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/8xknWag-SkI/5OsIhZWdDgAJ , so far we can not provide an automatic way to determine when will the reshuffle method is faster than ctas and when the reverse, and even in the case that reshuffle is faster the performance gain does not meet our expectation, for example 30% faster when expanding from 7 segments to 8 segments (without indexes, without vacuum).  Considering the complexity of the reshuffle method and the potential bugs it might cause (as the one in https://github.com/greenplum-db/gpdb/pull/7091#discussion_r262069344) we think it's not production ready yet.  In the future we might add it back if we can make it faster and stabler.

However, what makes reshuffle slower than expectation?  According to our experiments we think the deletion is the bottleneck, unless we could make deletion fast enough we could not get the benefit from reshuffle method in average cases.

**Comment by ashwinstar** [comment]
On Wed, Mar 6, 2019 at 9:03 PM Ning Yu <notifications@github.com> wrote:

> PR is missing to provide clear justification of why the reshuffle method
> came out as looser. Wish to understand the retirement is due to performance
> reasons or implementation risk. Tried to search mailing list but failed to
> find clear rationals for this there either.
>
> Are we retiring reshuffle just to avoid the risk of new implementation and
> continue to use the tried and tested method? or ignoring that fact
> benchmarks proved its a concept worth not visiting at all in future if we
> had no time pressure?
>
> Please can you post benchmarks numbers to clarify why reshuffle method
> didn't help. General cases of expansion would be going from 100 segments to
> 110 maybe, doesn't it prove to be not helpful for such cases? Reading full
> data and hash calculation for all the visible tuples is consistent between
> reshuffle and ctas. Reshuffle adds marking buffer dirty for deletes which
> ctas doesn't but ctas always has full rewrite of data. But for AO / CO
> tables reshuffle might still be winner over ctas as only should be touching
> visimap. Understand reshuffle can generate more work for future vacuum but
> still should help to get expand performed faster which was one of the
> goals. Where exactly is the bottleneck for reshuffle?
>
> @ashwinstar <https://github.com/ashwinstar> previous discussion and some
> benchmark data can be found here:
> https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/8xknWag-SkI/5OsIhZWdDgAJ
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__groups.google.com_a_greenplum.org_d_msg_gpdb-2Ddev_8xknWag-2DSkI_5OsIhZWdDgAJ&d=DwMFaQ&c=lnl9vOaLMzsy2niBC8-h_K-7QJuNJEsFrzdndhuJ3Sw&r=gxIaqms7ncm0pvqXLI_xjkgwSStxAET2rnZQpzba2KM&m=kPcj8kVqwMEcSVpSvVOU17mNGhSE-HjZ-AV8vViusGQ&s=3U8Z4RHKuSCHtkI7-UTdzeveVp2Wn4v5teW-h7lxN6Y&e=>
> , so far we can not provide an automatic way to determine when will the
> reshuffle method is faster than ctas and when the reverse, and even in the
> case that reshuffle is faster the performance gain does not meet our
> expectation, for example 30% faster when expanding from 7 segments to 8
> segments (without indexes, without vacuum).
>
Yes, had read through that mail thread, but it misses to provide concrete
benchmark numbers to come to any conclusion. Expand from 7 to 8 sounds toy
usecase and we should really evaluate based on more practical usecases when
bench-marking for gpexpand.

Considering the complexity of the reshuffle method and the potential bugs
> it might cause (as the one in #7091 (comment)
> <https://github.com/greenplum-db/gpdb/pull/7091#discussion_r262069344>)
> we think it's not production ready yet. In the future we might add it back
> if we can make it faster and stabler.
>
Sure, that helps. So, it seems we are retiring more out of risk and not
performance. This should be communicated and tracked in commit message when
retiring to help our future selves.

> However, what makes reshuffle slower than expectation? According to our
> experiments we think the deletion is the bottleneck, unless we could make
> deletion fast enough we could not get the benefit from reshuffle method in
> average cases.
>
Hmmmm, Thanks, good to know. Having this bottleneck also captured as part
of commit message will be helpful.

**Comment by pivotal-ning-yu** [comment]
On Fri, Mar 8, 2019 at 1:45 AM Ashwin Agrawal <notifications@github.com>
wrote:

> On Wed, Mar 6, 2019 at 9:03 PM Ning Yu <notifications@github.com> wrote:
>
> > PR is missing to provide clear justification of why the reshuffle method
> > came out as looser. Wish to understand the retirement is due to
> performance
> > reasons or implementation risk. Tried to search mailing list but failed
> to
> > find clear rationals for this there either.
> >
> > Are we retiring reshuffle just to avoid the risk of new implementation
> and
> > continue to use the tried and tested method? or ignoring that fact
> > benchmarks proved its a concept worth not visiting at all in future if we
> > had no time pressure?
> >
> > Please can you post benchmarks numbers to clarify why reshuffle method
> > didn't help. General cases of expansion would be going from 100 segments
> to
> > 110 maybe, doesn't it prove to be not helpful for such cases? Reading
> full
> > data and hash calculation for all the visible tuples is consistent
> between
> > reshuffle and ctas. Reshuffle adds marking buffer dirty for deletes which
> > ctas doesn't but ctas always has full rewrite of data. But for AO / CO
> > tables reshuffle might still be winner over ctas as only should be
> touching
> > visimap. Understand reshuffle can generate more work for future vacuum
> but
> > still should help to get expand performed faster which was one of the
> > goals. Where exactly is the bottleneck for reshuffle?
> >
> > @ashwinstar <https://github.com/ashwinstar> previous discussion and some
> > benchmark data can be found here:
> >
> https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/8xknWag-SkI/5OsIhZWdDgAJ
> > <
> https://urldefense.proofpoint.com/v2/url?u=https-3A__groups.google.com_a_greenplum.org_d_msg_gpdb-2Ddev_8xknWag-2DSkI_5OsIhZWdDgAJ&d=DwMFaQ&c=lnl9vOaLMzsy2niBC8-h_K-7QJuNJEsFrzdndhuJ3Sw&r=gxIaqms7ncm0pvqXLI_xjkgwSStxAET2rnZQpzba2KM&m=kPcj8kVqwMEcSVpSvVOU17mNGhSE-HjZ-AV8vViusGQ&s=3U8Z4RHKuSCHtkI7-UTdzeveVp2Wn4v5teW-h7lxN6Y&e=
> >
> > , so far we can not provide an automatic way to determine when will the
> > reshuffle method is faster than ctas and when the reverse, and even in
> the
> > case that reshuffle is faster the performance gain does not meet our
> > expectation, for example 30% faster when expanding from 7 segments to 8
> > segments (without indexes, without vacuum).
> >
> Yes, had read through that mail thread, but it misses to provide concrete
> benchmark numbers to come to any conclusion. Expand from 7 to 8 sounds toy
> usecase and we should really evaluate based on more practical usecases when
> bench-marking for gpexpand.
>
> Considering the complexity of the reshuffle method and the potential bugs
> > it might cause (as the one in #7091 (comment)
> > <https://github.com/greenplum-db/gpdb/pull/7091#discussion_r262069344>)
> > we think it's not production ready yet. In the future we might add it
> back
> > if we can make it faster and stabler.
> >
> Sure, that helps. So, it seems we are retiring more out of risk and not
> performance. This should be communicated and tracked in commit message when
> retiring to help our future selves.
>
> > However, what makes reshuffle slower than expectation? According to our
> > experiments we think the deletion is the bottleneck, unless we could make
> > deletion fast enough we could not get the benefit from reshuffle method
> in
> > average cases.
> >
> Hmmmm, Thanks, good to know. Having this bottleneck also captured as part
> of commit message will be helpful.
>

Thanks for the suggestion, I'll update the commit messages.

**Comment by pivotal-ning-yu** [comment]
Updated the commit message and the PR description.

**Comment by ashwinstar** [comment]
@pivotal-ning-yu Thanks a lot for patiently dealing with my questions and incorporating the comments. Message seems very good to me now. :+1: we are ready to have this PR in.

**Comment by pivotal-ning-yu** [comment]
Thanks for the reviews, we'll merge this PR on master now.

**Pull request merged**